File: src/main/java/core/procedures/cons/Append.java
package core.procedures.cons;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Cons;
import core.scm.Type;

public final class Append extends AFn {

  public Append() {
    super(new FnArgsBuilder().rest(Type.ProperList.class).last(Object.class).build());
  }

  @Override
  public String getName() {
    return "append";
  }

  @Override
  public Object apply(Object... args) {
    Object result = Cons.EMPTY;
    for (Object arg : args) {
      result = append(result, arg);
    }
    return result;
  }

  public static Object append(Object first, Object second) {
    if (Cons.isNull(first)) {
      return second;
    }
    return Cons.cons(Car.car(first), append(Cdr.cdr(first), second));
  }
}

File: src/main/java/core/procedures/cons/Append.kt
package core.procedures.cons

import core.procedures.AFn
import core.scm.Cons
import core.scm.Type

class Append : AFn<Any?, Any?>(name = "append", restArgsType = Type.ProperList::class.java, lastArgType = Any::class.java) {

    private val car = Car()
    private val cdr = Cdr()

    override operator fun invoke(args: Array<out Any?>) = args.fold(Cons.EMPTY as Any?, this::invoke)

    override operator fun invoke(arg1: Any?, arg2: Any?): Any? = when {
        Cons.isPair(arg1) -> Cons.cons(car(arg1), invoke(cdr(arg1), arg2))
        else -> arg2
    }
}
